pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+            } else {
+                XcmWeights::<T>::kill();
+            }
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Returns the interior multilocation for this container chain para id. This is a relative
+        /// multilocation that can be used in the `descend_origin` XCM opcode.
+        pub fn interior_multilocation(para_id: ParaId) -> InteriorMultiLocation {
+            let container_chain_account = T::GetParathreadAccountId::convert(para_id);
+            let account_junction = Junction::AccountId32 {
+                id: container_chain_account,
+                network: None,
+            };
+
+            InteriorMultiLocation::X1(account_junction)
+        }
+
+        /// Returns a multilocation that can be used in the `deposit_asset` XCM opcode.
+        /// The `interior_multilocation` can be obtained using `Self::interior_multilocation`.
+        pub fn absolute_multilocation(
+            interior_multilocation: InteriorMultiLocation,
+        ) -> MultiLocation {
+            let mut l = interior_multilocation;
+            l.push_front(Junction::Parachain(T::SelfParaId::get().into()))
+                .expect("multilocation too long");
+            MultiLocation::from(l)
+        }
+
+        fn on_collator_instantaneous_core_requested(para_id: ParaId) -> DispatchResult {
+            let mut in_flight_orders = InFlightOrders::<T>::get();
+            if in_flight_orders.contains(&para_id) {
+                return Err(Error::<T>::OrderAlreadyExists.into());
+            }
+            in_flight_orders
+                .try_insert(para_id)
+                .map_err(|_| Error::<T>::InFlightLimitReached)?;
+
+            // Check that the para id is a parathread
+            let parathread_params = T::GetParathreadParams::get_parathread_params(para_id);
+            if parathread_params.is_none() {
+                return Err(Error::<T>::NotAParathread.into());
+            }
+
+            // TODO: also compare the latest slot from pallet_author_noting with parathread_params.slot_frequency
+
+            let xcm_weights_storage =
+                XcmWeights::<T>::get().ok_or(Error::<T>::XcmWeightStorageNotSet)?;
+
+            let withdraw_amount = xcm_weights_storage.buy_execution_cost;
+
+            // Send xcm to the relay
+            // Use a derivative account from the sovereign account based on the paraId
+            // Buy on-demand cores
+            // Any failure should return everything to the derivative account
+
+            // Don't use utility::as_derivative because that will make the tanssi sovereign account
+            // pay for fees, instead use `DescendOrigin` to make the container chain sovereign account
+            // pay for fees. The container chain sovereign account is derived from the tanssi sovereign
+            // account.
+            // TODO: when coretime is implemented, buy core there instead of buying on-demand cores
```

Comment:
```suggestion
            // TODO: when coretime is implemented, use coretime instantaneous credits instead of buying on-demand cores at the price defined by the relay
```